aiorospy/src/aiorospy/action.py

```diff
@@ -28,3 +35,120 @@ def __init__(self, ns, action_spec, execute, loop=None):
+
+    async def feedback(self):
+        while True:
+            terminal_status = asyncio.create_task(self._done_event.wait())
+            new_feedback = asyncio.create_task(self._feedback_q.async_q.get())
+            done, pending = await asyncio.wait(
+                {terminal_status, new_feedback},
+                return_when=asyncio.FIRST_COMPLETED)
+            if new_feedback in done:
+                yield new_feedback.result()
+            elif terminal_status in done:
+                return
+            else:
+                raise RuntimeError("Unexpected termination condition")
+
+    async def wait_for_status(self, status):
+        while True:
+            await self._status_event.wait()
+            if status in self._old_statuses:
+                return
+            elif self._done_event.is_set():
+                raise RuntimeError(f"Action is done, will never reach status {status}")
+
+    async def done(self):
+        return self._done_event.wait()
+
+    def cancel(self):
+        # This gets injected by AsyncActionClient after init
+        raise NotImplementedError()
+
+    def _transition_cb(self, goal_handle):
+        self._transition_q.sync_q.put((
+            goal_handle.get_goal_status(),
+            goal_handle.get_comm_state(),
+            goal_handle.get_result()
+        ))
+
+    def _feedback_cb(self, goal_handle, feedback):
+        self._feedback_q.sync_q.put(feedback)
+
+    async def _process_transitions(self):
+        while not self._done_event.is_set():
+            status, comm_state, result = await self._transition_q.async_q.get()
+            self._status_event.set()
+
+            if status not in self._old_statuses:
+                self.status = status
+                self._old_statuses.add(status)
+
+            if comm_state == CommState.DONE:
+                self.result = result
+                self._done_event.set()
+            else:
+                self._status_event.clear()
+
+
+class AsyncActionClient:
+
+    def __init__(self, name, action_spec, loop=None):
+        self._loop = loop if loop is not None else asyncio.get_running_loop()
+
+        self.name = name
+        self._client = ActionClient(name, action_spec)
+        self._status_sub = AsyncSubscriber(name + "/status", GoalStatusArray, loop=self._loop, queue_size=1)
+
+    def send_goal(self, goal):
+        async_handle = AsyncGoalHandle(loop=self._loop)
+        sync_handle = self._client.send_goal(
+            goal,
+            transition_cb=async_handle._transition_cb,
+            feedback_cb=async_handle._feedback_cb,
+        )
+        async_handle.cancel = sync_handle.cancel
+        return async_handle
+
+    async def wait_for_server(self):
+        while True:
+            status_message = await self._status_sub.get()
+
+            # (pbovbel) the below replicates the behavior in actionlib.ActionClient.wait_for_server
+            server_id = status_message._connection_header['callerid']
```

Comment:
https://www.python.org/dev/peps/pep-0008/#string-quotes

I dunno if I'm being a zealot but the consistency really helps when we want to start using the other quote inside strings.  I just follow PEP8s "just pick one and stick with it"